src/field/field_factory.cxx

```diff
@@ -99,30 +99,29 @@ FieldFactory::~FieldFactory() {
 
 }
 
-const Field2D FieldFactory::create2D(const std::string &value, const Options *opt,
-                                     Mesh *localmesh, CELL_LOC loc,
-                                     BoutReal t) {
+const Field2D FieldFactory::create2D(const std::string& value, const Options* opt,
+                                     Mesh* localmesh, CELL_LOC loc, BoutReal t) {
+  return create2D(parse(value, opt), localmesh, loc, t);
+}
 
-  if(localmesh == nullptr)
+const Field2D FieldFactory::create2D(FieldGeneratorPtr gen, Mesh* localmesh, CELL_LOC loc,
+                                     BoutReal t) {
+  AUTO_TRACE();
+  
+  if (localmesh == nullptr)
     localmesh = fieldmesh;
-  if(localmesh == nullptr)
+  if (localmesh == nullptr)
```

Comment:
This double `localmesh` check looks a bit odd -- I can see what it is doing (second check checks `fieldmesh/=nullptr` essentially) but this is a bit unusual in terms of what we do elsewhere so it might be good to add a comment. Also the exception message could perhaps be a little more descriptive (to note what we've tried, say). Should we finally try to fall back to global mesh before deciding to throw?

Reply:
Yes, it's a bit of an odd pattern, but I think we do occasionally use this elsewhere (not that that's an argument in favour of it!).

We normally fall back to the global mesh in the ctor, which is probably the best place to do so.

Does anyone have an opinion on which of these is more readable:

```cpp
if ((localmesh == nullptr) and (fieldmesh == nullptr)) {
  throw BoutException("No valid mesh found: FieldFactory not created with mesh and no mesh passed in");
}

if (localmesh == nullptr) {
  localmesh = fieldmesh;
}
```
or:
```cpp
if (localmesh == nullptr) {
  if (fieldmesh == nullptr) {
    throw BoutException("No valid mesh found: FieldFactory not created with mesh and no mesh passed in");
  } else {
    localmesh = fieldmesh;
  }
}
```

The first makes the precondition more explicit, while the second avoids the repeated check.

Reply:
I find the second one easier to read. I think that's because when I look at the first one it makes me think about the values of `localmesh` and `fieldmesh` at the same time, whereas in the second I think about the state of `localmesh` first, and then about `fieldmesh`.

Reply:
I'll make these changes in separate PR, as they're orthogonal to this one